### DISCUSSION TITLE
Move equivalent of dynamic dispatch?

### DISCUSSION THREAD

**Discussion by lumos42** [aptos-labs/aptos-developer-discussions#322] [Questions]
I understand that Move doesn't support dynamic dispatch like Solidity. Is there any Move idiomatic way to achieve the same behaviour?

We have 2 use cases:
1. In our protocol each pool will have one of many possible interest rate models - like fixed interest and linear interest model. In Solidity, I would just save the interest rate model's contract address, and since all models are following a particular interface, we could just call a function on the saved address. The only way to do this in Move I can think of is to have an if-else condition every time we need to determine the interest.
2. Is there any way to write a proxy function in Move - a function which accepts the contract address, function name/identifier, and arguments and makes the call? In Solidity it would be written like this:
```solidity
function exec(address target, uint256 value, bytes calldata data) external {
    // pre-checks
    (bool success,) = target.call{ value: value }(data);
    // post-checks
}
```

**Comment by 0x-j** [comment]
[You can add transfer hook to the fungible asset](https://aptos.dev/en/build/smart-contracts/fungible-asset#dispatchable-fungible-asset-advanced), but that's the only dynamic dispatch we can do on aptos today. But let me ask around.

**Comment by MoonShiesty** [comment]
1. you'd need to write a switch table or invert the dependency between protocol and interest model modules so each interest model module calls into protocol
the move-v2 compiler will support other approaches, but those are the only two i can think of right now

2. no proxies but the aptos framework has an idiomatic way to upgrade code

**Reply by lumos42** [reply]
> no proxies but the aptos framework has an idiomatic way to upgrade code

I meant proxy functions, not UpgaradableProxy contracts. But since posting this question, I've learnt that such a thing is not possible with move.

Where can I read more about move-v2 compiler?

**Reply by MoonShiesty** [reply]
https://medium.com/aptoslabs/the-future-of-move-at-aptos-17d0656dcc31

move v2 will support generalized higher-order functions, which i believe would be clean solution to the problem you have

**Comment by kshitijc1506** [comment]
Have a look at [this twitter post](https://x.com/david_wolinsky/status/1816163534209966095). Might be helpful.
